encoder/libav_encoder.cpp

```diff
@@ -499,27 +492,101 @@ void LibAvEncoder::encode(AVPacket *pkt, unsigned int stream_id)
 		// Rescale from the codec timebase to the stream timebase.
 		av_packet_rescale_ts(pkt, codec_ctx_[stream_id]->time_base, out_fmt_ctx_->streams[stream_id]->time_base);
 
+		// When we are paused, keep track of how long passes in the pause state. Needed to make video continuous
+		if (!write_frames_)
+			pause_duration_ += current_timestamp_ - previous_video_timestamp_;
+		previous_video_timestamp_ = current_timestamp_;
 		std::scoped_lock<std::mutex> lock(output_mutex_);
-		// pkt is now blank (av_interleaved_write_frame() takes ownership of
-		// its contents and resets pkt), so that no unreferencing is necessary.
-		// This would be different if one used av_write_frame().
-		ret = av_interleaved_write_frame(out_fmt_ctx_, pkt);
+
+		// We need a copy of what the packet timestamp was
+		int64_t packet_ts_ = pkt->pts;
+
+		/*
+		Need to detect what state the packet came from
+		If the timestamp is after a pause timestamp and before an unpause ts then it must be paused, and hence thrown away
+		Additionally, the timestamp can be after a pause event but the current unpause timestamp is equal to zero, and hence is also paused
+		if the timestamp is after an unpause timestamp and after the latest pause timestamp then it must be written (unpaused state)
+
+		In the following, the logic all happens when we are passed a video frame. i.e. starting a new segment, initating split.
+		Audio just follows a through afterwards. This is needed to make sure that the first thing in the video file is a keyframe.
+		*/
+
+		if (options_->keypress)
+		{
+			// Alter the timestamps to be continuous
+			pkt->pts -= pause_duration_;
+			pkt->dts -= pause_duration_;
+			if (packet_ts_ > pause_timestamp_ && (packet_ts_ < unpause_timestamp_ || !unpause_timestamp_))
+			{
+				// This must be a paused frame. We shall discard it
+				if (!first_frame_ && stream_id == Video)
+				{
+					// Upon pausing, set variables up for next segment, and if we are splitting, call function to create new output
+					first_frame_ = true;
+					if (options_->split)
+						nextSegment();
+				}
+				write_frames_ = false;
+			}
+
+			if (packet_ts_ >= unpause_timestamp_ && packet_ts_ > pause_timestamp_ && unpause_timestamp_ != 0)
+			{
+				// This frame fits into the 'unpaused' category
+				if (first_frame_ && pkt->flags & AV_PKT_FLAG_KEY && stream_id == Video)
+				{
+					// This is the first frame that is a keyframe. This will be the start of a new segment/split.
+					ret = av_interleaved_write_frame(out_fmt_ctx_, pkt);
+					first_frame_ = false;
+					write_frames_ = true;
+				}
+				// If there's nothing special going on, just write the frame
+				else if (!first_frame_ && write_frames_)
+					ret = av_interleaved_write_frame(out_fmt_ctx_, pkt);
+			}
+		}
+
+		if (options_->segment)
```

Comment:
``options_->segment`` probably wants to now became our new fancy ``TimeVal`` type so users can specify units on the command line.